nexus-and-alo/nexus-strawman.md

```diff
@@ -0,0 +1,361 @@
+
+1. User asks the Nexus Service author for the endpoint of the Nexus Service Instance they are trying to consume. For Temporal based Nexus Service Instances, the endpoint will not be HTTP or gRPC but rather a custom uri formulation.
+2. User registers provided endpoint with their SDK to avoid hardcoding the endpoint into all of their application logic. 
+3. User calls the external Nexus Service using a Workflow-specific Nexus API which is provided by Temporal.
+    ```go
+    // calling a Nexus from a Temporal Workflow
+    func MyWorkflow(ctx workflow.Context, name string) (string, error) {
+        client := temporalnexus.NewAloClient(temporalnexus.AloClientOptions{Endpoint: "temporal:72aA64A0jhOL99B2"})
+        // Start some ALO
+        alo, err := client.StartAlo(ctx, temporalnexus.StartAloOptions{
+            Name: "foo",
+            ID:   "my-id",
+            Arg:  name,
+        })
+        if err != nil {
+            return "", err
+        }
+        // Workflow implementation...
+    }
+    ```
+
+## Concepts
+
+### Nexus Instantiation
+
+Instantiating a Nexus Service makes it possible for the Nexus Backend to receive traffic meant for that specific Instance. Nexus Instantiation does make your Nexus Service Instance endpoint(s) available for traffic, but until you register and run corresponding Nexus Task Workers, no forward progress will be made on incoming requests. In this way, sending requests at this stage is similar to starting Workflows without any Workflow Workers running.
+
+**Registration:**
+
+For the sake of this proposal, we will only consider the case where Nexus is registered via `nexus` CLI. In the future we may evaluate and consider alternative methods of registering a Nexus Service Instance. Registering a Nexus Service Instance would be done with the following command structure:
+
+`nexus service register --name my-service --description "My service" --backend nexus-backend.example.com` 
+
+This would result in the following output being displayed
+
+// TODO (Chad): if this spits out a Temporal uri the cli should probably have an option to denote that
+
+```bash
+$ nexus service register --service-name my-service --service-description "My service" --backend nexus-backend.example.com
+
+Nexus registered successfully with instance name "my-service"!
+Available at:
+	Temporal: temporal:72aA64A0jhOL99B2
+```
+
+**What is the behavior after running this command?**
+
+![Nexus behavior](./images/behavior.png)
+
+The Nexus Instance is now registered and available via Nexus backend and should be reachable via the provided endpoint. Sending requests to the endpoint is possible even at this stage, but there should be no expectation of a response or of forward progress of the request. This is because while the Nexus Instance is registered and available, we have not run any corresponding Nexus Workers. In this way, sending requests at this stage is similar to starting Workflows without any Workflow Workers running.
```

Comment:
Requires more information about the "temporal" protocol, is it not a sync request / response protocol?
If so, how is this possible?

Reply:
I don't have the details, but I believe @mfateev wants requests to be able to survive when all workers are down. Of course Nexus is not Temporal specific, so it's more like "backend does whatever it needs to with requests to ensure delivery".

Reply:
Agree with Chad but can elaborate.